smslibrary/src/main/java/com/eis/smslibrary/SMSPeer.java

```diff
@@ -0,0 +1,152 @@
+
+/**
+ * Implementation of {@link Peer} for SMS communication channel.
+ *
+ * @author Luca Crema, Marco Mariotto.
+ */
+public class SMSPeer implements Peer<String>, Comparable<SMSPeer> {
+
+    public static final int MAX_TELEPHONE_NUMBER_LENGTH = 20;
+    public static final int MIN_TELEPHONE_NUMBER_LENGTH = 3;
+
+    private String telephoneNumber;
+
+    /**
+     * @param telephoneNumber Address for the peer.
+     * @throws InvalidTelephoneNumberException If telephoneNumber check is not {@link TelephoneNumberState#TELEPHONE_NUMBER_VALID}.
+     */
+    public SMSPeer(String telephoneNumber) throws InvalidTelephoneNumberException{
+        TelephoneNumberState telephoneNumberState = SMSPeer.checkPhoneNumber(telephoneNumber);
+        if(telephoneNumberState != TelephoneNumberState.TELEPHONE_NUMBER_VALID)
+            throw new InvalidTelephoneNumberException(telephoneNumberState);
+
+        this.telephoneNumber = telephoneNumber;
+    }
+
+    /**
+     * Checks if the given phone number is valid.
+     *
+     * @param telephoneNumber the phone number to check.
+     * @return The {@link TelephoneNumberState} of the telephone number after the tests.
+     */
+    public static TelephoneNumberState checkPhoneNumber(String telephoneNumber) {
+        //Check if the number is shorter than the MAX.
+        if (telephoneNumber.length() > MAX_TELEPHONE_NUMBER_LENGTH) {
+            return TelephoneNumberState.TELEPHONE_NUMBER_TOO_LONG;
+        }
+        //Check if the number is longer than the MIN.
+        if (telephoneNumber.length() < MIN_TELEPHONE_NUMBER_LENGTH) {
+            return TelephoneNumberState.TELEPHONE_NUMBER_TOO_SHORT;
+        }
+        //Check if it's actually a number and doesn't contain anything else
+        //First we have to remove the "+"
+        if (!telephoneNumber.substring(1, telephoneNumber.length() - 1).matches("[0-9]+")) {
+            return TelephoneNumberState.TELEPHONE_NUMBER_NOT_A_NUMBER;
+        }
+        //Check if there is a country code.
+        if (telephoneNumber.charAt(0) != '+') {
+            return TelephoneNumberState.TELEPHONE_NUMBER_NO_COUNTRY_CODE;
+        }
+        //If it passed all the tests we are sure the number is valid.
+        return TelephoneNumberState.TELEPHONE_NUMBER_VALID;
+    }
+
+    /**
+     * Compares this object with the specified object for order.
+     * @param o The object to be compared.
+     * @return A negative integer, zero, or a positive integer as this object is less than, equal to, or greater than the specified object.
+     */
+    @Override
+    public int compareTo(SMSPeer o) {
+        return telephoneNumber.compareTo(o.telephoneNumber);
+    }
+
+    /**
+     * Retrieves the peer's address.
+     * @return Peer's address.
+     */
+    @Override
+    public String getAddress() {
+        return telephoneNumber;
+    }
+
+    /**
+     * Sets or updates the peer's address.
+     * @param telephoneNumber New address for the peer.
+     */
```

Comment:
i think it’s better to block an address setting if it could turn the whole peer in a not valid state
In this case the method could return the state

Reply:
I was thinking about removing the whole setter, it's not useful and probably violates some encapsulation rules